mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.hive;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hive.service.cli.CLIService;
+import org.apache.hive.service.cli.HiveSQLException;
+import org.apache.hive.service.cli.OperationHandle;
+import org.apache.hive.service.cli.RowSet;
+import org.apache.hive.service.cli.SessionHandle;
+import org.apache.hive.service.cli.session.HiveSession;
+import org.apache.hive.service.server.HiveServer2;
+import org.apache.iceberg.common.DynMethods;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkState;
+
+/**
+ * Test class for running HiveQL queries, essentially acting like a Beeline shell in tests.
+ *
+ * It takes a metastore URL via conf, and spins up an HS2 instance which connects to it. The shell will only accept
+ * queries if it has been previously initialized via {@link #start()}, and a session has been opened via
+ * {@link #openSession()}. Prior to calling {@link #start()}, the shell should first be configured with props that apply
+ * across all test cases by calling {@link #setHiveConfValue(String, String)} ()}. On the other hand, session-level conf
+ * can be applied anytime via {@link #setHiveSessionValue(String, String)} ()}, once we've opened an active session.
+ */
+public class TestHiveShell {
```

Comment:
Do you know how the metastore connections are handled in HS2? Does each HS2 service have a connection pool? Or are connections created per session?

Why did you choose to use a session per test case? Is that to avoid some sort of state leak? Are sessions not supposed to be used concurrently?

Reply:
> Do you know how the metastore connections are handled in HS2? 

During query compilation, HS2 calls the Metastore via thrift (using the `ThriftHiveMetastore$Client`), synchronously in the same thread where the Driver performs the other compilation tasks. The created metastore client is cached and reused throughout the life of the Driver execution.

> Does each HS2 service have a connection pool?

The HS2 service we're using for submitting queries (`CLIClient`) does not use a thread pool (it only uses a thread pool when async compilation/execution is enabled, which isn't in our tests or by default).

> Why did you choose to use a session per test case? Is that to avoid some sort of state leak? 

Exactly, to avoid state/configuration leaks, in order to provide a 'clean slate' for each test method. Test methods could therefore go and set their session-level variables individually that they need for their test scenarios, without risks of test config spilling over (which could be hard to debug).

Reply:
From your answers to the first two, it sounds like there will be one connection per driver that is reused, and planning/execution is single-threaded so we shouldn't have multiple connections from a single test case.

And if the driver or its underlying connection to HMS is reused across sessions, then this should open a single connection to the Hive MetaStore per test suite. Is that correct?

If so, then I think we should be able to turn down the metastore pool size from 15 to 5 because this fixes the problem where a new connection is used for each command.

Reply:
The metastore client is cached in threadlocal, so each test method will create one and reuse it across its query executions (so not 1 connection per test suite but 1 per test method). However, since the connections are properly closed by calling `session.close()` at the end of each test method, I see no reason to keep the higher pool size. Changed it back to the default size.